sharktank/sharktank/ops/sharded_impls.py

```diff
@@ -1711,14 +1711,44 @@ def topk_split(
         indices_split = SplitPrimitiveTensor(ts=indices, shard_dim=input.shard_dim)
         return values_split, indices_split
     else:
-        # TODO: Implement more efficient version which does a
-        #       topk on each shard before transferring and then adjusts
-        #       the indices.
-        gathered = sharded_cat(input)
-        values, indices = topk(gathered, k=k, dim=dim, largest=largest, sorted=sorted)
-        values = replicate(values, count=input.shard_count, devices=input.devices)
-        indices = replicate(indices, count=input.shard_count, devices=input.devices)
-        return values, indices
+        # TODO: implement using all_reduce_topk when IREE supports it
+
+        all_pair = []
+        offset = 0
+        running_vals, running_inds = None, None
+        for i, shard in enumerate(input.shards):
+            v_loc, i_loc = topk(shard, k=k, dim=dim, largest=largest, sorted=sorted)
+
+            i_glob = i_loc + offset
+            offset += shard.shape[dim]
+
+            if i != 0:
+                v_loc = transfer_to_logical_device(v_loc, input.devices[0])
+                i_glob = transfer_to_logical_device(i_glob, input.devices[0])
+
+            else:
+                v_loc = barrier_on_logical_device(v_loc, input.devices[0])
+                i_glob = barrier_on_logical_device(i_glob, input.devices[0])
+
+            if running_vals is None:
+                running_vals, running_inds = v_loc, i_glob
+            else:
+                cand_vals = torch.cat([running_vals, v_loc], dim=dim)
+                cand_inds = torch.cat([running_inds, i_glob], dim=dim)
+
+                running_vals, pos = torch.topk(
+                    cand_vals, k=k, dim=dim, largest=largest, sorted=sorted
+                )
+                running_inds = torch.take_along_dim(cand_inds, pos, dim=dim)
```

Comment:
I'm pretty sure this is going to create one large serial dependency chain between the different iterations, which I don't think IREE will break for us.

I think the following would be faster:
1. topk each shard in parallel
2. adjust the indices to be indices into full tensor rather than into each shard.
3. Transfer values and indices to devices[0].
4. Cat values together, same with indices.
5. Perform 1 last topk on values and use resulting indices to index into the adjusted index tensor from step 3.
6. Return